chainerx_cc/chainerx/context.cc

```diff
@@ -265,6 +255,22 @@ std::vector<BackpropId> Context::GetInnerBackpropIds(const BackpropId& backprop_
     return inner_backprop_ids;
 }
 
+Backend& Context::RegisterBackend(const std::string& backend_name, std::unique_ptr<Backend, context_detail::BackendDeleter> backend) {
+    // In a multi-threaded case, backends_[backend_name] may already exist at this point.
+    // In that case, the backend created in `Context::GetBackend` is thrown away.
+    Backend& backend_ref = *backend;
+    {
+        std::lock_guard<std::mutex> lock{mutex_};
+        auto pair = backends_.emplace(backend_name, std::move(backend));
+        // Initialize the backend only if emplaced.
+        if (!pair.second) {
+            return backend_ref;
+        }
+    }
+    backend_ref.Initialize();
```

Comment:
`Initialize` should be called within the lock. Otherwise the second thread could receive a backend while it's being initialized.

Reply:
Fixed.